pkg/controller/common/api/v1/common.go

```diff
@@ -0,0 +1,158 @@
+package v1
+
+import (
+	"fmt"
+	"strings"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// RelativeObjectMap holds object related to given parent object.
+// The structure is [GroupVersionKind] -> [common.relativeName] -> *unstructured.Unstructured
+// where:
+//   GroupVersionKind - identifies type stored in entry
+//   relativeName() - return path to object in relation to parent
+//   *unstructured.Unstructured - object to store
+type RelativeObjectMap map[GroupVersionKind]map[string]*unstructured.Unstructured
+
+// InitGroup initializes a map for given schema.GroupVersionKind if not yet initialized
+func (m RelativeObjectMap) InitGroup(gvk schema.GroupVersionKind) {
+	internalGvk := GroupVersionKind{gvk}
+	if m[internalGvk] == nil {
+		m[internalGvk] = make(map[string]*unstructured.Unstructured)
+	}
+}
+
+// Insert inserts given obj to RelativeObjectMap regarding parent
+func (m RelativeObjectMap) Insert(parent v1.Object, obj *unstructured.Unstructured) {
+	internalGvk := GroupVersionKind{obj.GroupVersionKind()}
+	if m[internalGvk] == nil {
+		m.InitGroup(obj.GroupVersionKind())
+	}
+	group := m[internalGvk]
+	name := relativeName(parent, obj)
+	group[name] = obj
+}
+
+// InsertAll inserts given slice of objects to RelativeObjectMap regarding parent
+func (m RelativeObjectMap) InsertAll(parent v1.Object, objects []*unstructured.Unstructured) {
+	for _, object := range objects {
+		m.Insert(parent, object)
+	}
+}
+
+// FindGroupKindName search object by name in each schema.GroupKind (ignoring version part)
+func (m RelativeObjectMap) FindGroupKindName(gk schema.GroupKind, name string) *unstructured.Unstructured {
+	// The map is keyed by group-version-kind, but we don't know the version.
+	// So, check inside any GVK that matches the group and kind, ignoring version.
+	for key, objects := range m {
+		if key.GroupKind() == gk {
+			for n, object := range objects {
+				if n == name {
+					return object
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// relativeName returns the name of the object relative to the parent.
+// If the parent is cluster scoped and the object namespaced scoped the
+// name is of the format <namespace>/<name>. Otherwise the name of the object
+// is returned.
+func relativeName(parent v1.Object, obj *unstructured.Unstructured) string {
+	if parent.GetNamespace() == "" && obj.GetNamespace() != "" {
+		return fmt.Sprintf("%s/%s", obj.GetNamespace(), obj.GetName())
+	}
+	return obj.GetName()
+}
+
+// ReplaceObject replaces the object with the same name & namespace as
+// the given object with the contents of the given object. If no object exists
+// in the existing map then no action is taken.
+func (m RelativeObjectMap) ReplaceObject(parent v1.Object, obj *unstructured.Unstructured) {
```

Comment:
nit: Is it possible to rename this to MaybeReplaceObject or ReplaceObjectIfExists

Reply:
sure